pkg/operator/ceph/object/objectstore_test.go

```diff
@@ -815,3 +816,31 @@ func TestGetRealmKeyArgs(t *testing.T) {
 		assert.Equal(t, "", secret)
 	})
 }
+
+func TestUpdateZoneEndpointList(t *testing.T) {
+	type args struct {
+		zoneEndpointList []string
+		oldEndpoint      string
+		newEndpoint      string
+	}
+
+	tests := []struct {
+		name string
+		args args
+		want []string
+	}{
+		{"adding new endpoint to emptylist", args{zoneEndpointList: []string{}, oldEndpoint: "", newEndpoint: "http://new-endpoint"}, []string{"http://new-endpoint"}},
+		{"adding new endpoint to existing list", args{zoneEndpointList: []string{"http://rgw-endpoint-1"}, oldEndpoint: "", newEndpoint: "http://new-endpoint"}, []string{"http://rgw-endpoint-1", "http://new-endpoint"}},
+		{"updating same endpoint to list", args{zoneEndpointList: []string{"http://rgw-endpoint-1"}, oldEndpoint: "http://rgw-endpoint-1", newEndpoint: "http://rgw-endpoint-1"}, []string{"http://rgw-endpoint-1"}},
+		{"updating different endpoint to single endpoint list", args{zoneEndpointList: []string{"http://rgw-endpoint-1"}, oldEndpoint: "http://rgw-endpoint-1", newEndpoint: "http://rgw-endpoint-2"}, []string{"http://rgw-endpoint-2"}},
+		{"updating different endpoint to muliple endpoint list", args{zoneEndpointList: []string{"http://rgw-endpoint-1", "http://rgw-endpoint-2"}, oldEndpoint: "http://rgw-endpoint-2", newEndpoint: "http://rgw-endpoint-3"}, []string{"http://rgw-endpoint-1", "http://rgw-endpoint-3"}},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := updateZoneEndpointList(tt.args.zoneEndpointList, tt.args.oldEndpoint, tt.args.newEndpoint); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("updateZoneEndpointList() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```

Comment:
There is at least one test missing here:
1. the newEndpoint already exists in the middle of the existing list, and the oldEndpoint does not equal the newEndpoint

I think you'll find that your implementation fails this test.

Also consider what other tests may be missing. I proposed 6 tests in the go playground link I sent, and there are 5 here. You also include some tests that I missed wherin (1) the oldEndpoint is the empty string and wherein (2) oldEndpoint and newEndpoint are equal.

My tests were divided into two sections with minimal input lists and longer input lists. To continue that trend and also include empty string and equal version, I think this means there should be at least 10 unit tests.

Reply:
Writing good unit tests is hard, but it is important to take the time and effort to fully exercise the inputs of functions and fully understand the desired outputs. I prototyped this again on the go playground and wrote 16 tests. I also discovered a bugs in my initial implementation relating to an empty input, and I found some behavior that I didn't like w.r.t. desired/undesired inputs being equal.

This is a hint that I should at least see 16 tests that are equivalent to what I prototyped in this PR. And add whatever other tests I may have missed or that you deem necessary.